File: Python Codes/Lista Dinâmica Fila/dynamic_list_queue.py
import random


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def create_node(data):
    return Node(data)


def enqueue(front, rear, data):
    new_node = create_node(data)
    if rear[0] is None:
        # Se a fila está vazia
        front[0] = new_node
        rear[0] = new_node
    else:
        # Adiciona no final da fila
        rear[0].next = new_node
        rear[0] = new_node


def dequeue(front, rear):
    if front[0] is None:
        print("A fila está vazia, não há elementos para remover.")
        return

    front[0] = front[0].next

    if front[0] is None:
        # Se a fila ficou vazia, o rear também deve ser None
        rear[0] = None

    print("Elemento removido do início da fila.")


def print_queue(front):
    current = front[0]
    while current is not None:
        print(f"{current.data} -> ", end='')
        current = current.next
    print("NULL")


def print_menu_and_queue(front):
    if front[0] is not None:
        print("Fila Dinâmica: ", end='')
        print_queue(front)
    print("\n1. Inserir na fila\n2. Remover da fila\n3. Sair\nEscolha uma opção: ", end='')


def main():
    front = [None]  # Início da fila
    rear = [None]   # Final da fila

    while True:
        print_menu_and_queue(front)
        choice = int(input())

        if choice == 1:
            data = random.randint(0, 99)
            enqueue(front, rear, data)
            print(f"Número {data} enfileirado.")
        elif choice == 2:
            dequeue(front, rear)
        elif choice == 3:
            # Libera a memória alocada para a fila
            while front[0] is not None:
                front[0] = front[0].next
            break
        else:
            print("Opção inválida.")


if __name__ == "__main__":
    main()

File: Códigos/C_Codes/Lista Dinâmica Fila/dynamicListQueue.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <locale.h>

// Definição do nó da fila
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Função para criar um novo nó
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Erro ao alocar memória.\n");
        exit(1); // Encerra o programa em caso de falha na alocação
    }
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Função para enfileirar (inserir no final)
void enqueue(Node** front, Node** rear, int data) {
    Node* newNode = createNode(data);
    if (*rear == NULL) {
        // Se a fila está vazia
        *front = *rear = newNode;
    } else {
        // Adiciona no final da fila
        (*rear)->next = newNode;
        *rear = newNode;
    }
}

// Função para desenfileirar (remover do início)
void dequeue(Node** front, Node** rear) {
    if (*front == NULL) {
        printf("A fila está vazia, não há elementos para remover.\n");
        return;
    }

    Node* temp = *front;
    *front = (*front)->next;

    if (*front == NULL) {
        // Se a fila ficou vazia, o rear também deve ser NULL
        *rear = NULL;
    }

    free(temp);
    printf("Elemento removido do início da fila.\n");
}

// Função para imprimir a fila
void printQueue(Node* front) {
    Node* current = front;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

int main() {
    setlocale(LC_ALL, "Portuguese");
    Node* front = NULL; // Início da fila
    Node* rear = NULL;  // Final da fila
    int choice, data;

    srand(time(NULL));

    while (1) {
        printf("\nFila Dinâmica: ");
        printQueue(front);
        printf("\n1. Inserir na fila\n");

        if (front != NULL) {
            printf("2. Remover da fila\n");
        }

        printf("3. Sair\nEscolha uma opção: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                data = rand() % 100;
                enqueue(&front, &rear, data);
                printf("Número %d enfileirado.\n", data);
                break;
            case 2:
                if (front != NULL) {
                    dequeue(&front, &rear);
                } else {
                    printf("A fila está vazia, não há elementos para remover.\n");
                }
                break;
            case 3:
                // Libera a memória alocada para a fila
                while (front != NULL) {
                    Node* temp = front;
                    front = front->next;
                    free(temp);
                }
                return 0;
            default:
                printf("Opção inválida.\n");
        }
    }
}
